File: client/go/src/example/connect/test/cli/asyncMenu.go
package cli

import (
	"fmt"
	"strconv"
	"strings"
	"example/cli"
	"biostar/service/connect"
)

var (
	asyncMenuItems []*cli.MenuItem
)

func init() {
	asyncMenuItems = []*cli.MenuItem{
		&cli.MenuItem{
			"1", "Add async connections", addAsyncConnection, false,
		},
		&cli.MenuItem{
			"2", "Delete async connections", deleteAsyncConnection, false,
		},
		&cli.MenuItem{
			"3", "Refresh the connection list", showAsyncConnection, false,
		},		
		&cli.MenuItem{
			"q", "Return to Main Menu", nil, true,
		},
	}
}

func showAsyncMenu() error {
	err := showAsyncConnection()

	if err != nil {
		return err
	}

	cli.ShowMenu("Async Menu", asyncMenuItems)
	return nil
}


func showAsyncConnection() error {
	devList, err := connectSvc.GetDeviceList()

	if err != nil {
		newErr := fmt.Errorf("Cannot get the device list: %v", err)
		fmt.Println(newErr)
		return newErr
	}

	asyncList := []*connect.DeviceInfo{}

	for _, device := range devList {
		if device.AutoReconnect {
			asyncList = append(asyncList, device)
		}
	}

	fmt.Printf("\n***** Async connections: %v\n", len(asyncList))

	for _, device := range asyncList {
		fmt.Printf("%v\n", device)
	}

	return nil
}


func getAsyncConnectInfo() ([]*connect.AsyncConnectInfo, error) {
	connInfos := []*connect.AsyncConnectInfo{}

	for true {
		connInfo := &connect.AsyncConnectInfo{}

		var devIDStr string
	
		inputs := []*cli.UserInput{
			&cli.UserInput{
				"Enter the device ID (Press just ENTER if no more device)", "", &devIDStr,
			},
		}

		err := cli.GetUserInput(inputs)
	
		if err != nil {
			return nil, err
		}	
	
		if strings.TrimSpace(devIDStr) == "" {
			break
		}

		devID, err := strconv.Atoi(devIDStr)

		if err != nil  {
			newErr := fmt.Errorf("Invalid device ID: %v", err)
			fmt.Println(newErr)
			return nil, newErr
		}

		connInfo.DeviceID = uint32(devID)
	
		syncConnInfo, err := getConnectInfo()

		if err != nil {
			return nil, err
		}

		connInfo.IPAddr = syncConnInfo.IPAddr
		connInfo.Port = syncConnInfo.Port
		connInfo.UseSSL = syncConnInfo.UseSSL

		connInfos = append(connInfos, connInfo)
	}

	return connInfos, nil
}


func addAsyncConnection() error {
	connInfos, err := getAsyncConnectInfo()

	if err != nil {
		return err
	}

	if len(connInfos) == 0 {
		err := fmt.Errorf("No device information")
		fmt.Println(err)
		return err
	}

	fmt.Printf("Adding asynchronous connections for %d devices...\n", len(connInfos))

	err = connectSvc.AddAsyncConnection(connInfos)

	if err != nil {
		return err
	}

	showAsyncConnection()

	return nil
}


func deleteAsyncConnection() error {
	fmt.Printf("\nEnter the device IDs to delete\n")

	devIDs, err := cli.GetDeviceIDs()

	if err != nil {
		return err
	}

	fmt.Printf("Deleting asynchronous connections for %v devices...\n", len(devIDs))

	err = connectSvc.DeleteAsyncConnection(devIDs)

	if err != nil {
		return err
	}

	showAsyncConnection()

	return nil
}

File: client/node/example/connect/test/cli/asyncMenu.js
const menu = require('../../../cli/menu');
const connect = require('../../../connect');
const mainMenu = require('./mainMenu');
const readlineSync = require('readline-sync');

async function showAsyncConnection() {
  var connInfos = [];

  try {
    var devList = await connect.getDeviceList();

    for(let i = 0; i < devList.length; i++) {
      if(devList[i].autoreconnect) {
        connInfos.push(devList[i]);
      }
    }

    console.log(`\n***** Async connections: ${connInfos.length}`);
    console.log(connInfos);

    return connInfos;
  }  catch (err) {
    console.error('Cannot get the async connections: ', err);
    return null;
  }
}

function getAsyncConnectInfo() {
  var connInfos = [];

  var userInputs = [];

  userInputs.push(new menu.UserInput('Enter the IP address of the device', ''));
  userInputs.push(new menu.UserInput('Enter the port of the device (default: 51211)', '51211'));
  userInputs.push(new menu.UserInput('Use SSL y/n (default: n)', 'n'));

  while(true) {
    var devIDStr = readlineSync.question('>> Enter the device ID (Press just ENTER if no more device): ');

    if(devIDStr.trim() == '') {
      break
    }

    var devID = Number(devIDStr);

    if(isNaN(devID)) {
      console.error('Invalid device ID: ', devIDStr);
      break;
    }

    var inputs = menu.getUserInput(userInputs);

    var connInfo = new connect.connectMessage.AsyncConnectInfo();
    connInfo.setDeviceid(devID)
    connInfo.setIpaddr(inputs[0]);
    
    var port = Number(inputs[1]);
    if(isNaN(port)) {
      console.error('Invalid port number: ', inputs[1]);
      break;
    }

    connInfo.setPort(port);
    inputs[2].trim();

    if(inputs[2] == 'y' || inputs[2] == 'Y') {
      connInfo.setUsessl(true);
    } else {
      connInfo.setUsessl(false);
    }

    connInfos.push(connInfo);
  }

  return connInfos;
}

async function addAsyncConnection() {
  var connInfos = getAsyncConnectInfo();

  if(connInfos.length == 0) {
    console.error('No async connection to add');
    return;
  }

	console.log(`Adding asynchronous connections for ${connInfos.length} devices...`);

  try {
    await connect.addAsyncConnection(connInfos);
    await showAsyncConnection();
  } catch (err) {
    console.error('Cannot add the async connections: ', err);
  }
}


async function deleteAsyncConnection() {
  var deviceIDs = menu.getDeviceIDs();

  if(deviceIDs.length == 0) {
    console.error('No async connection to delete');
    return;
  }

	console.log(`Deleting asynchronous connections for ${deviceIDs.length} devices...`);

  try {
    await connect.deleteAsyncConnection(deviceIDs);
    await showAsyncConnection();
  } catch (err) {
    console.error('Cannot delete the async connections: ', err);
  }
}


async function showMenu() {
  await showAsyncConnection();

  var menuItems = [];
  
  menuItems.push(new menu.MenuItem('1', 'Add async connections', addAsyncConnection, false));
  menuItems.push(new menu.MenuItem('2', 'Delete async connections', deleteAsyncConnection, false));
  menuItems.push(new menu.MenuItem('3', 'Refresh the connection list', showAsyncConnection, false));
  menuItems.push(new menu.MenuItem('q', 'Return to Main Menu', mainMenu.showMenu, true));

  menu.showMenu('Async Menu', menuItems)  
}

module.exports.showMenu = showMenu;


